server/rpc.go

```diff
@@ -50,6 +51,64 @@ type Broadcaster interface {
 	Sign([]byte) ([]byte, error)
 }
 
+type BroadcastSessionsManager struct {
+	broadcastSessions []*BroadcastSession
+	broadcastSessMap  map[string]*BroadcastSession
+	sessLock          *sync.Mutex
+}
+
+func (bsm *BroadcastSessionsManager) selectSession() *BroadcastSession {
+	bsm.sessLock.Lock()
+	defer bsm.sessLock.Unlock()
+
+	// minOrchs is maximum number of inflight sessions
+	minOrchs := int(HTTPTimeout / SegLen)
+	if len(bsm.broadcastSessions) < minOrchs {
+		bsm.checkStatus()
```

Comment:
Note a small complication here.

* We really want the checkStatus call to be async because it could take a while.
* We need to ensure proper locking within checkStatus, especially where we access and mutate the existing list of sessions.
* Lastly, "checkStatus" is starting to seem like an odd name now. Perhaps a more accurate naming would be "refreshSessions" or some such.

This would work:
```golang
func (bsm *BroadcastSessionsManager) selectSession() *BroadcastSession {
  // minOrchs is maximum number of inflight sessions
  minOrchs := int(HTTPTimeout / SegLen)
  bsm.sessLock.Lock()
  numSess = len(bsm.broadcastSessions) // we use this a bunch of other places
  if numSess < minOrchs {
    bsm.sessLock.Unlock()
    go bsm.refreshSessions()
  }
  defer bsm.sessLock.Unlock()
  ... rest of function here ...
}

func (bsm *BroadcastSessionsManager) refreshSessions() {
  newBroadcastSessions := ...
  if (len(newBroadcastSessions) <= 0) {
    return
  }
  uniqueSessions := make([]*BroadcastSession, 0, len(newBroadcastSessions))
  bsm.sessLock.Lock()
  defer bsm.sessLock.Unlock()
  ... rest of function here ... 
}
```

Reply:
Ah I see what you're doing there. I misunderstood locks a bit here - I thought they were "carried along" to every function that was called within another function. I see now that we need to use a lock again in `refreshSessions`. `refreshSessions` works for me!

Reply:
>  I misunderstood locks a bit here - I thought they were "carried along" to every function that was called within another function

That is correct -- locks are "carried through" until unlocked somewhere. (The unlocking doesn't necessarily have to be in the same part of the call stack or even the same thread, but this naturally makes things difficult to reason about, so it's preferable to keep locking/unlocking localized where possible.)

The reason we need to lock within refreshSessions is because it's being called asynchronously, and we can't count on the lock from `selectSession` to fully "carry through". For example:

```
selectSession : Acquire lock
refreshSessions : Start execution
selectSession : Release lock
refreshSessions : Stop execution
```

So here we need to lock in both selectSessions and refreshSessions in order to serialize accesses to the set of sessions.

Reply:
Oh I see. That makes more sense! Thanks for the explanation Josh